File: ContiguousArray_LC525.java
// Time Complexity : O(n)
// Space Complexity :O(n)
// Did this code successfully run on Leetcode :Yes

class Solution {
    public int findMaxLength(int[] nums) {
       int count=0;
        HashMap<Integer,Integer> map = new HashMap<>();
        int max=0;
        int len =0;
        
        //to detect the initial balance sub array
        map.put(0,-1);
        
        for(int i=0;i<nums.length;i++){
            if(nums[i]==1){
                count++;
            }
            else {
                count--;
            }
            
            if(map.containsKey(count)){
                len = i - map.get(count);
                max = Math.max(max,len);
            }
            else{
                map.put(count, i);
            }
        }
        
        return max;
    }
}

File: contiguous-array_LC_525.py
#TC:O(N) SC:O(N)
#Running runningSum

class Solution:
    def findMaxLength(self, nums: List[int]) -> int:
        dict = {}
        dict[0] = -1
        runningSum=0
        maxSubArray = 0

        for i in range(len(nums)):
          
            if nums[i]==1:
                runningSum = runningSum+1
            else: 
                runningSum = runningSum-1
             
            
            if runningSum in dict.keys(): 
                maxSubArray = max(maxSubArray, i - dict[runningSum])
                
            else:
                dict[runningSum] = i
           
        
        return maxSubArray
